src/Serialization.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -18,51 +29,37 @@ public GraphQLRequestDeserializer(Action<JsonSerializerOptions> configure)
 
         public async Task<GraphQLRequestDeserializationResult> FromBodyAsync(Stream stream, long? contentLength)
         {
-            int length;
-            byte[] jsonBytes;
             var sharedArrayPool = ArrayPool<byte>.Shared;
-            bool wasRented = false;
-            if (contentLength.HasValue)
+            if (contentLength.HasValue && contentLength < MaxSharedBufferSize)
             {
-                length = (int)contentLength;
-                jsonBytes = sharedArrayPool.Rent(length);
-                wasRented = true;
+                var length = (int)contentLength.Value; // MaxSharedBufferSize < int.MaxValue so this is OK
+                var jsonBytes = sharedArrayPool.Rent(length); 
                 await stream.ReadAsync(jsonBytes, 0, length);
-            }
-            else
-            {
-                using var ms = new MemoryStream();
-                await stream.CopyToAsync(ms).ConfigureAwait(false);
-                if (!ms.TryGetBuffer(out var streamBytes))
+
+                try
                 {
-                    return new GraphQLRequestDeserializationResult();
+                    return Process(jsonBytes.AsSpan(0, length));
                 }
-                jsonBytes = streamBytes.Array;
-                length = streamBytes.Count;
-            }
-
-            try
-            {
-                return Process(jsonBytes, length);
-            }
-            finally
-            {
-                if (wasRented)
+                finally
                 {
                     sharedArrayPool.Return(jsonBytes);
                 }
             }
+            else
+            {
+                using var ms = new MemoryStream();
```

Comment:
Is there any reason you need to buffer here for System.Text.Json? It supports streaming de-serialization.

Reply:
We have this weird case where a request could be a JSON object or and array, (single graph operation or a batch of ops), so we're kinda forced to read the first byte to the do `GetTokenType` (in same class) so we know whether to deserialize into T or T[]. Open to any suggestions you have. Just reading up on PipeReader, and thinking, like in the [docs here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/request-response?view=aspnetcore-3.1), could we just pipe read the first byte, check for `(byte)'['` or `(byte)'{'`, bail out early if neither, then somehow just pass the stream straight to System.Text.Json?

Reply:
https://gist.github.com/davidfowl/058b607055257a52c9ea9185a3efff71

Reply:
Wow! I did not know that it is possible.
` reader.AdvanceTo(buffer.Start, buffer.Start);`

Reply:
This is the analog of `Peek` that we were looking for!

Reply:
The one remaining challenge is, `.BodyReader` is netcoreapp3+, so the interface that needed to support `netstandard20` as well gets tricky.

Reply:
Perhaps the interface just has the method accept `HttpRequest` since the SystemTextJson project is targeted at only 3.0 and then it can just access the property off the request.

Reply:
OK